homeassistant/components/zha/cover.py

```diff
@@ -0,0 +1,176 @@
+):
+    """Set up the ZHA covers."""
+    entities = []
+    for discovery_info in discovery_infos:
+        zha_dev = discovery_info["zha_device"]
+        channels = discovery_info["channels"]
+
+        entity = ZHA_ENTITIES.get_entity(DOMAIN, zha_dev, channels, ZhaCover)
+        if entity:
+            entities.append(entity(**discovery_info))
+
+    if entities:
+        async_add_entities(entities, update_before_add=True)
+
+
+@STRICT_MATCH(channel_names=CHANNEL_COVER)
+class ZhaCover(ZhaEntity, CoverDevice):
+    """Representation of a ZHA cover."""
+
+    def __init__(self, unique_id, zha_device, channels, **kwargs):
+        """Init this sensor."""
+        super().__init__(unique_id, zha_device, channels, **kwargs)
+        self._cover_channel = self.cluster_channels.get(CHANNEL_COVER)
+        self._current_position = None
+
+    async def async_added_to_hass(self):
+        """Run when about to be added to hass."""
+        await super().async_added_to_hass()
+        await self.async_accept_signal(
+            self._cover_channel, SIGNAL_ATTR_UPDATED, self.async_set_position
+        )
+
+    @callback
+    def async_restore_last_state(self, last_state):
+        """Restore previous state."""
+        self._state = last_state.state
+        if "current_position" in last_state.attributes:
+            self._current_position = last_state.attributes["current_position"]
+
+    @property
+    def is_closed(self):
+        """Return if the cover is closed."""
+        if self.current_cover_position is None:
+            return None
+        return self.current_cover_position == 0
+
+    @property
+    def current_cover_position(self):
+        """Return the current position of ZHA cover.
+
+        None is unknown, 0 is closed, 100 is fully open.
+        """
+        return self._current_position
+
+    def async_set_position(self, pos):
+        """Handle position update from channel."""
+        _LOGGER.debug("setting position: %s", pos)
+        self._current_position = 100 - pos
+        if self._current_position == 0:
+            self._state = STATE_CLOSED
+        elif self._current_position == 100:
+            self._state = STATE_OPEN
+        self.async_schedule_update_ha_state()
+
+    def async_set_state(self, state):
+        """Handle state update from channel."""
+        _LOGGER.debug("state=%s", state)
+        self._state = state
+        self.async_schedule_update_ha_state()
+
+    async def async_open_cover(self, **kwargs):
+        """Open the window cover."""
+        res = await self._cover_channel.up_open()
+        if isinstance(res, list) and res[1] is Status.SUCCESS:
+            self.async_set_state(STATE_OPENING)
+
+    async def async_close_cover(self, **kwargs):
+        """Close the window cover."""
+        res = await self._cover_channel.down_close()
+        if isinstance(res, list) and res[1] is Status.SUCCESS:
+            self.async_set_state(STATE_CLOSING)
```

Comment:
With this info we can implement `is_closing` cover entity property.

https://github.com/home-assistant/home-assistant/blob/4dc39492a57c505c1915e679f06198204458763c/homeassistant/components/cover/__init__.py#L225-L228